File: frontend/src/lib/humanGate.test.ts
import { describe, it, expect } from 'vitest'
import { isHumanGateLabel, findLabel } from './humanGate'
import type { Workflow, WorkflowLabel, WorkflowTransition } from '../api/client'

function label(name: string, opts: { agentIgnore?: boolean; isTerminal?: boolean } = {}): WorkflowLabel {
  return {
    id: name,
    workflow_id: 'wf',
    name,
    color: '#000000',
    sort_order: 0,
    agent_ignore: opts.agentIgnore ? 1 : 0,
    is_terminal: opts.isTerminal ? 1 : 0,
  }
}

function transition(
  from: string,
  to: string,
  trigger: WorkflowTransition['trigger_type'] = 'agent',
): WorkflowTransition {
  return { id: `${from}->${to}`, workflow_id: 'wf', from_label: from, to_label: to, trigger_type: trigger }
}

function workflow(labels: WorkflowLabel[], transitions: WorkflowTransition[]): Workflow {
  return {
    id: 'wf',
    name: 'wf',
    description: '',
    labels,
    transitions,
    created_at: '',
    updated_at: '',
  }
}

describe('isHumanGateLabel', () => {
  it('returns false when workflow is undefined', () => {
    expect(isHumanGateLabel(undefined, 'review')).toBe(false)
  })

  it('returns false when the label does not exist in the workflow', () => {
    const wf = workflow([label('a')], [])
    expect(isHumanGateLabel(wf, 'missing')).toBe(false)
  })

  it('is a gate when agent_ignore is set, regardless of transitions', () => {
    const wf = workflow(
      [label('review', { agentIgnore: true })],
      [transition('review', 'done', 'agent')],
    )
    expect(isHumanGateLabel(wf, 'review')).toBe(true)
  })

  it('is a gate when every outgoing transition is human-only', () => {
    const wf = workflow(
      [label('review'), label('done', { isTerminal: true })],
      [transition('review', 'done', 'human')],
    )
    expect(isHumanGateLabel(wf, 'review')).toBe(true)
  })

  it('is not a gate when an outgoing transition is agent or both', () => {
    const wfAgent = workflow(
      [label('review'), label('done', { isTerminal: true })],
      [transition('review', 'x', 'human'), transition('review', 'done', 'agent')],
    )
    expect(isHumanGateLabel(wfAgent, 'review')).toBe(false)

    const wfBoth = workflow(
      [label('review'), label('done', { isTerminal: true })],
      [transition('review', 'done', 'both')],
    )
    expect(isHumanGateLabel(wfBoth, 'review')).toBe(false)
  })

  it('is not a gate for a terminal label even with no outgoing transitions', () => {
    const wf = workflow([label('done', { isTerminal: true })], [])
    expect(isHumanGateLabel(wf, 'done')).toBe(false)
  })

  it('is not a gate for a non-terminal label with zero outgoing transitions (dead end)', () => {
    const wf = workflow([label('stuck')], [])
    expect(isHumanGateLabel(wf, 'stuck')).toBe(false)
  })
})

describe('findLabel', () => {
  it('finds a label by name', () => {
    const wf = workflow([label('review')], [])
    expect(findLabel(wf, 'review')?.name).toBe('review')
  })

  it('returns undefined when workflow is undefined or label missing', () => {
    expect(findLabel(undefined, 'review')).toBeUndefined()
    const wf = workflow([label('review')], [])
    expect(findLabel(wf, 'missing')).toBeUndefined()
  })
})

File: backend/internal/workflow/humangate_test.go
package workflow_test

import (
	"testing"

	"github.com/myinisjap/agent-task-editor/backend/internal/storage/gen"
	"github.com/myinisjap/agent-task-editor/backend/internal/workflow"
)

func transition(from, to, trigger string) gen.WorkflowTransition {
	return gen.WorkflowTransition{FromLabel: from, ToLabel: to, TriggerType: trigger}
}

func gateLabel(name string, isTerminal, agentIgnore int64) gen.WorkflowLabel {
	return gen.WorkflowLabel{Name: name, IsTerminal: isTerminal, AgentIgnore: agentIgnore}
}

func TestIsHumanGateLabel(t *testing.T) {
	cases := []struct {
		name        string
		labels      []gen.WorkflowLabel
		transitions []gen.WorkflowTransition
		labelName   string
		want        bool
	}{
		{
			name:      "unknown label",
			labels:    []gen.WorkflowLabel{gateLabel("todo", 0, 0)},
			labelName: "missing",
			want:      false,
		},
		{
			name:   "terminal label is never a gate",
			labels: []gen.WorkflowLabel{gateLabel("done", 1, 0)},
			transitions: []gen.WorkflowTransition{
				transition("done", "archived", "human"),
			},
			labelName: "done",
			want:      false,
		},
		{
			name:      "agent_ignore is always a gate, even with no transitions",
			labels:    []gen.WorkflowLabel{gateLabel("not_ready", 0, 1)},
			labelName: "not_ready",
			want:      true,
		},
		{
			name:   "all outgoing transitions human is a gate",
			labels: []gen.WorkflowLabel{gateLabel("review", 0, 0)},
			transitions: []gen.WorkflowTransition{
				transition("review", "approved", "human"),
				transition("review", "rejected", "human"),
			},
			labelName: "review",
			want:      true,
		},
		{
			name:   "mixed human/agent outgoing is not a gate",
			labels: []gen.WorkflowLabel{gateLabel("review", 0, 0)},
			transitions: []gen.WorkflowTransition{
				transition("review", "approved", "human"),
				transition("review", "retry", "agent"),
			},
			labelName: "review",
			want:      false,
		},
		{
			name:   "all agent outgoing is not a gate",
			labels: []gen.WorkflowLabel{gateLabel("in_progress", 0, 0)},
			transitions: []gen.WorkflowTransition{
				transition("in_progress", "done", "agent"),
			},
			labelName: "in_progress",
			want:      false,
		},
		{
			name:      "zero outgoing transitions is not a gate",
			labels:    []gen.WorkflowLabel{gateLabel("dead_end", 0, 0)},
			labelName: "dead_end",
			want:      false,
		},
	}

	for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {
			got := workflow.IsHumanGateLabel(c.labels, c.transitions, c.labelName)
			if got != c.want {
				t.Errorf("IsHumanGateLabel(%q) = %v, want %v", c.labelName, got, c.want)
			}
		})
	}
}
